fdbclient/SpecialKeySpace.actor.cpp

```diff
@@ -0,0 +1,377 @@
+		if (endIter->value() != nullptr) wait(endIter->value()->normalizeKeySelectorActor(endIter->value(), ryw, &end));
+		end.offset < 1 ? --endIter : ++endIter;
+	}
+
+	actualEndOffset = end.offset;
+	if (endIter == pks->impls.ranges().begin())
+		end.setKey(pks->range.begin);
+	else if (endIter == pks->impls.ranges().end())
+		end.setKey(pks->range.end);
+
+	if (!end.isFirstGreaterOrEqual()) {
+		// The Key Selector points to key outside the whole special key space
+		TraceEvent(SevInfo, "EndKeySelectorPointsOutside")
+		    .detail("TerminateKey", end.getKey())
+		    .detail("TerminateOffset", end.offset);
+		if (end.offset < 1 && endIter == pks->impls.ranges().begin())
+			result.readToBegin = true;
+		else
+			result.readThroughEnd = true;
+		end.offset = 1;
+	}
+	// Handle all corner cases like what RYW does
+	// return if range inverted
+	if (actualBeginOffset >= actualEndOffset && begin.getKey() >= end.getKey()) {
+		TEST(true);
+		return RangeResultRef(false, false);
+	}
+	// If touches begin or end, return with readToBegin and readThroughEnd flags
+	if (beginIter == pks->impls.ranges().end() || endIter == pks->impls.ranges().begin()) {
+		TEST(true);
+		return result;
+	}
+	state RangeMap<Key, SpecialKeyRangeBaseImpl*, KeyRangeRef>::Ranges ranges =
+	    pks->impls.intersectingRanges(KeyRangeRef(begin.getKey(), end.getKey()));
+	// TODO : workaround to write this two together to make the code compact
+	// The issue here is boost::iterator_range<> doest not provide rbegin(), rend()
+	iter = reverse ? ranges.end() : ranges.begin();
+	if (reverse) {
+		while (iter != ranges.begin()) {
+			--iter;
+			if (iter->value() == nullptr) continue;
+			KeyRangeRef kr = iter->range();
+			KeyRef keyStart = kr.contains(begin.getKey()) ? begin.getKey() : kr.begin;
+			KeyRef keyEnd = kr.contains(end.getKey()) ? end.getKey() : kr.end;
+			Standalone<RangeResultRef> pairs = wait(iter->value()->getRange(ryw, KeyRangeRef(keyStart, keyEnd)));
+			result.arena().dependsOn(pairs.arena());
+			// limits handler
+			for (int i = pairs.size() - 1; i >= 0; --i) {
+				result.push_back(result.arena(), pairs[i]);
+				// Note : behavior here is even the last k-v pair makes total bytes larger than specified, it's still
+				// returned. In other words, the total size of the returned value (less the last entry) will be less
+				// than byteLimit
+				limits.decrement(pairs[i]);
+				if (limits.isReached()) {
+					result.more = true;
+					result.readToBegin = false;
+					return result;
+				};
+			}
+		}
+	} else {
+		for (iter = ranges.begin(); iter != ranges.end(); ++iter) {
+			if (iter->value() == nullptr) continue;
+			KeyRangeRef kr = iter->range();
+			KeyRef keyStart = kr.contains(begin.getKey()) ? begin.getKey() : kr.begin;
+			KeyRef keyEnd = kr.contains(end.getKey()) ? end.getKey() : kr.end;
+			Standalone<RangeResultRef> pairs = wait(iter->value()->getRange(ryw, KeyRangeRef(keyStart, keyEnd)));
```

Comment:
If getRagne on one key space throws error, will user gets the error?
If it blocks indefinitely, how will user know?

(Acknowledgement: This is from my discussion with @ajbeamon.)

Reply:
My personal understanding is that errors are thrown by underlying functions.
Thus the user will get the error that is thrown by the function directly.
What error to give back is depends on each function.

For the time issue, I think for in some scenarios like reading conflicting keys they are just reading local data, which should always be fine. For a function that needs to wait, like `status/json`, the underlying function implementation should handle the time issue. 
From the framework's perspective, I feel it is hard to unify the time issue here.